packages/common/src/api/tan-query/developerApps.ts

```diff
@@ -0,0 +1,215 @@
+      const sdk = await audiusSdk()
+
+      const { apiKey, apiSecret } = await sdk.developerApps.createDeveloperApp({
+        name,
+        description,
+        imageUrl,
+        userId: encodedUserId
+      })
+
+      await sdk.grants.createGrant({
+        userId: encodedUserId,
+        appApiKey: apiKey
+      })
+
+      return { name, description, imageUrl, apiKey, apiSecret }
+    }
+  })
+}
+
+export const useEditDeveloperApp = () => {
+  const { audiusSdk } = useAudiusQueryContext()
+  const queryClient = useQueryClient()
+
+  return useMutation({
+    mutationFn: async (args: UseEditDeveloperAppArgs) => {
+      const { name, description, imageUrl, userId, apiKey } = args
+      const encodedUserId = Id.parse(userId)
+      const sdk = await audiusSdk()
+
+      await sdk.developerApps.updateDeveloperApp({
+        appApiKey: apiKey,
+        name,
+        description,
+        imageUrl,
+        userId: encodedUserId
+      })
+
+      return { name, description, imageUrl, apiKey }
+    },
+    onMutate: (args) => {
+      const { userId, ...editedApp } = args
+
+      queryClient.invalidateQueries({
+        queryKey: [QUERY_KEYS.developerApps, userId]
+      })
+
+      const previousApps: DeveloperApp[] =
+        queryClient.getQueryData([QUERY_KEYS.developerApps, userId]) ?? []
+      const appIndex = previousApps.findIndex(
+        (app) => app.apiKey === editedApp.apiKey
+      )
+
+      const newApps = cloneDeep(previousApps)
+      newApps[appIndex] = {
+        ...previousApps[appIndex],
+        ...editedApp
+      }
+
+      queryClient.setQueryData([QUERY_KEYS.developerApps, userId], newApps)
+
+      // Return context with the previous apps
+      return { previousApps }
+    },
+    onError: (_error, args, context) => {
+      queryClient.setQueryData(
+        [QUERY_KEYS.developerApps, args.userId],
+        context?.previousApps
+      )
+    }
+  })
+}
+
+export const useDeleteDeveloperApp = () => {
+  const { audiusSdk } = useAudiusQueryContext()
+  const queryClient = useQueryClient()
+
+  return useMutation({
+    mutationFn: async (args: UseDeleteDeveloperAppArgs) => {
+      const { apiKey, userId } = args
+      const sdk = await audiusSdk()
+
+      await sdk.developerApps.deleteDeveloperApp({
+        userId: Id.parse(userId),
+        appApiKey: apiKey
+      })
+    },
+    onMutate: (args) => {
+      const { apiKey, userId } = args
+
+      queryClient.invalidateQueries({
+        queryKey: [QUERY_KEYS.developerApps, userId]
+      })
+
+      const previousApps: DeveloperApp[] | undefined = queryClient.getQueryData(
+        [QUERY_KEYS.developerApps, userId]
+      )
+
+      if (previousApps === undefined) {
+        return {
+          previousApps: []
+        }
+      }
+
+      // Splice out the deleted app
+      const appIndex = previousApps?.findIndex((app) => app.apiKey === apiKey)
+      const newApps = cloneDeep(previousApps).splice(appIndex, 1)
```

Comment:
nice! so cloning is necessary?

Reply:
making a new version so that the old version is not mutated so that we can roll back in the event of an error